### PR TITLE
Forth

### LINKED ISSUE #23: Forth exercise

### PR DESCRIPTION
**Pull request by glennj**
Closes #23

### CODE REVIEW

File: exercises/practice/forth/.meta/example.awk [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+############################################################
+# stack functions
+
+function push(value) {
+    stack[++s] = value
+}
+
+function pop(    value) {
```
Comment (IsaacG) [inline exercises/practice/forth/.meta/example.awk:50]:
```suggestion
function pop(value) {
```

File: exercises/practice/forth/.meta/example.awk [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+function check_stack_size(n) {
+    if (n == 2 && s == 1) die("only one value on the stack")
+    if (s == 0)           die("empty stack")
+}
+
+############################################################
+# forth builtin operations
+
+function arithmetic(operation,    v1, v2) {
```
Comment (IsaacG) [inline exercises/practice/forth/.meta/example.awk:68]:
```suggestion
function arithmetic(operation, v1, v2) {
```

File: exercises/practice/forth/.meta/example.awk [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    check_stack_size(1)
+    push(peek())
+}
+
+function drop() {
+    check_stack_size(1)
+    pop()
+}
+
+function swap(    v1, v2) {
```
Comment (IsaacG) [inline exercises/practice/forth/.meta/example.awk:92]:
```suggestion
function swap(v1, v2) {
```

File: exercises/practice/forth/.meta/example.awk [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    check_stack_size(2)
+    v2 = pop()
+    v1 = pop()
+    push(v2)
+    push(v1)
+}
+
+function over(    v1, v2) {
```
Comment (IsaacG) [inline exercises/practice/forth/.meta/example.awk:100]:
```suggestion
function over(v1, v2) {
```

File: exercises/practice/forth/.meta/example.awk [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    v2 = pop()
+    v1 = peek()
+    push(v2)
+    push(v1)
+}
+
+############################################################
+# forth user-defined operations
+
+function record_macro(    i) {
```
Comment (IsaacG) [inline exercises/practice/forth/.meta/example.awk:111]:
```suggestion
function record_macro(i) {
```

File: exercises/practice/forth/.meta/example.awk [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+function join(array, start, end, sep,    result, i) {
+    sep = separator(sep)
+    result = array[start]
+    for (i = start + 1; i <= end; i++)
+        result = result sep array[i]
+    return result
+}
+
+# too bad awk doesn't provide an array holding the fields
+function join_fields(start, end, sep,    result, i) {
```
Comment (IsaacG) [inline exercises/practice/forth/.meta/example.awk:146]:
```suggestion
function join_fields(start, end, sep, result, i) {
```

File: exercises/practice/forth/.meta/example.awk [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    exit 1
+}
+
+function is_integer(str) {
+    return (str ~ /^[+-]?[[:digit:]]+$/)
+}
+
+# based on
+# https://www.gnu.org/software/gawk/manual/html_node/Join-Function.html 
+function join(array, start, end, sep,    result, i) {
```
Comment (IsaacG) [inline exercises/practice/forth/.meta/example.awk:137]:
```suggestion
function join(array, start, end, sep, result, i) {
```

**Review by IsaacG — APPROVED** [review]

File: exercises/practice/forth/.meta/example.awk (hunk 1 above)
Comment (glennj) [reply exercises/practice/forth/.meta/example.awk:50]: The extra whitespace is awk convention. The only concept of "local variables" awk has are function parameters.

From [Function Definition Syntax](https://www.gnu.org/software/gawk/manual/html_node/Definition-Syntax.html)
> Usually when you write a function, you know how many names you intend to use for arguments and how many you intend to use as local variables. It is conventional to place some extra space between the arguments and the local variables, in order to document how your function is supposed to be used.

**Pull request merged**